src/PairedRouting.php

```diff
@@ -0,0 +1,901 @@
+
+			// On frontend, continue support case where the query var has been (manually) set.
+			global $wp_the_query;
+			if (
+				$wp_the_query instanceof WP_Query
+				&&
+				false !== $wp_the_query->get( $slug, false )
+			) {
+				return true;
+			}
+
+			// When not in a frontend context (e.g. the Customizer), the query var is the only possibility.
+			if (
+				is_admin()
+				&&
+				isset( $_GET[ $slug ] ) // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			) {
+				return true;
+			}
+
+			$url = amp_get_current_url();
+		}
+		return $this->get_paired_url_structure()->has_endpoint( $url );
+	}
+
+	/**
+	 * Turn a given URL into a paired AMP URL.
+	 *
+	 * @param string $url URL.
+	 * @return string AMP URL.
+	 */
+	public function add_endpoint( $url ) {
+		return $this->get_paired_url_structure()->add_endpoint( $url );
+	}
+
+	/**
+	 * Remove the paired AMP endpoint from a given URL.
+	 *
+	 * @param string $url URL.
+	 * @return string URL with AMP stripped.
+	 */
+	public function remove_endpoint( $url ) {
+		return $this->get_paired_url_structure()->remove_endpoint( $url );
+	}
+
+	/**
+	 * Determine whether a custom paired URL structure is being used.
+	 *
+	 * @return bool Whether custom paired URL structure is used.
+	 */
+	public function has_custom_paired_url_structure() {
+		return has_filter( 'amp_custom_paired_url_structure' );
+	}
+
+	/**
+	 * Get paired URLs for all available structures.
+	 *
+	 * @param string $url URL.
+	 * @return array Paired URLs keyed by structure.
+	 */
+	public function get_all_structure_paired_urls( $url ) {
+		$paired_urls = [];
+		foreach ( self::PAIRED_URL_STRUCTURES as $structure_slug => $structure_class ) {
+			/** @var PairedUrlStructure $structure */
+			$structure = $this->injector->make( $structure_class );
+
+			$paired_urls[ $structure_slug ] = $structure->add_endpoint( $url );
+		}
+
+		if ( $this->has_custom_paired_url_structure() ) {
+			$paired_urls[ self::PAIRED_URL_STRUCTURE_CUSTOM ] = $this->add_endpoint( $url );
+		}
+
+		return $paired_urls;
+	}
+
+	/**
+	 * Get paired URL examples.
+	 *
+	 * @return array[] Keys are the structures, values are arrays of paired URLs using the structure.
+	 */
+	public function get_paired_url_examples() {
+		$supported_post_types     = AMP_Post_Type_Support::get_supported_post_types();
+		$hierarchical_post_types  = array_intersect(
+			$supported_post_types,
+			get_post_types( [ 'hierarchical' => true ] )
+		);
+		$chronological_post_types = array_intersect(
+			$supported_post_types,
+			get_post_types( [ 'hierarchical' => false ] )
+		);
+
+		$examples = [];
+		foreach ( [ $chronological_post_types, $hierarchical_post_types ] as $post_types ) {
+			if ( empty( $post_types ) ) {
+				continue;
+			}
+			$posts = get_posts(
+				[
+					'post_type'   => $post_types,
+					'post_status' => 'publish',
+				]
+			);
+			foreach ( $posts as $post ) {
+				if ( count( AMP_Post_Type_Support::get_support_errors( $post ) ) !== 0 ) {
```

Comment:
Just a note: No examples would be shown if there are no published posts for the queried post types, or if all posts for a certain post type had did not support AMP (both scenarios are unlikely, but possible).

Reply:
True. Worst case is just no examples would be shown:

![image](https://user-images.githubusercontent.com/134745/105307407-52972000-5b71-11eb-8830-816d5c028026.png)

But the examples are already somewhat redundant since the forms are indicated in the options already.